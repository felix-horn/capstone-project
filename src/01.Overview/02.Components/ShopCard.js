import { useHistory } from 'react-router-dom'
import PropTypes from 'prop-types'
import styled from 'styled-components/macro'

import ShopTitle from '../01.UI-Elements/ShopTitle'
import ListItem from '../01.UI-Elements/ListItem'
import CheckedItemsSummary from '../01.UI-Elements/CheckedItemsSummary'
import {
  getUncheckedItemIds,
  getQuantityCheckedItems,
  getShopTitle,
  getItemTitle,
} from '../../services/database.services'

ShopCard.propTypes = {
  shopId: PropTypes.string.isRequired,
  database: PropTypes.object.isRequired,
}

export default function ShopCard({ shopId, database }) {
  const history = useHistory()
  const uncheckedIds = getUncheckedItemIds(database, shopId)
  const quantityCheckedIds = getQuantityCheckedItems(database, shopId)
  const shopTitle = getShopTitle(database, shopId)

  return (
    <CardLayout onClick={navigateToShop} data-testid="shop-card">
      <ShopTitlePositioned title={shopTitle} />
      {uncheckedIds.map((id) => (
        <ListItem key={id} title={getItemTitle(database, id)} />
      ))}
      {quantityCheckedIds > 0 && (
        <CheckedItemsSummary quantityCheckedIds={quantityCheckedIds} />
      )}
    </CardLayout>
  )
  function navigateToShop() {
    history.push({
      pathname: `/shop/${shopTitle}`,
      state: { shopId },
    })
  }
}

const CardLayout = styled.div`
  border: var(--border);
  border-radius: 8px;
  padding: 15px;
  display: grid;
  gap: 5px;
`
const ShopTitlePositioned = styled(ShopTitle)`
  margin-bottom: 8px;
`
