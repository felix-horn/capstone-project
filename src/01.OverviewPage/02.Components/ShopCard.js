import PropTypes from 'prop-types'
import styled from 'styled-components/macro'
import ShopTitle from '../01.UI-Elements/ShopTitle'
import ListItem from '../01.UI-Elements/ListItem'

ShopCard.propTypes = {
  database: PropTypes.object.isRequired,
}

export default function ShopCard({ database }) {
  const uncheckedIds = database.items.allIds.filter(
    (id) => !database.items.byId[id].isChecked
  )
  return (
    <>
      <ShopCardStyled>
        <ShopTitle title="Penny" />
        {uncheckedIds.map((id) => {
          const { title } = database.items.byId[id]
          return <ListItem key={id} title={title} />
        })}
      </ShopCardStyled>
    </>
  )
}

const ShopCardStyled = styled.div`
  border: var(--border);
  border-radius: 10px;
  padding: 10px;
  display: grid;
`
