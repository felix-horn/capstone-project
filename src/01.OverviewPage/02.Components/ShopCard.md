```js padded
<ShopCard
shopId={'x'} 
database={{
    shops: {
    allIds: ['x'],
    byId: {
      x: {
        id: 'x',
        title: 'Shop X',
        items: ['a', 'b', 'c'],
      },
     },
  },
  items: {
    allIds: ['a', 'b', 'c'],
    byId: {
      a: {
        id: 'a',
        title: 'Item A',
        isChecked: false,
      },
      b: {
        id: 'b',
        title: 'Item B',
        isChecked: false,
      },
      c: {
        id: 'c',
        title: 'Item C',
        isChecked: true,
      },
    },
  },
}}
/>
```