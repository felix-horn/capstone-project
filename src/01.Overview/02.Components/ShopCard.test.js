import { render } from '@testing-library/react'
import shallow from 'react-test-renderer/shallow'
import user from '@testing-library/user-event'
import { BrowserRouter as Router } from 'react-router-dom'
import ShopCard from './ShopCard'

const testDatabase = {
  shops: {
    allIds: ['x', 'y'],
    byId: {
      x: {
        id: 'x',
        title: 'shopX',
        items: ['a', 'b'],
      },
      y: {
        id: 'y',
        title: 'shopY',
        items: ['c'],
      },
    },
  },
  items: {
    allIds: ['a', 'b', 'c'],
    byId: {
      a: {
        id: 'a',
        title: 'itemA',
        isChecked: false,
      },
      b: {
        id: 'b',
        title: 'itemB',
        isChecked: true,
      },
      c: {
        id: 'c',
        title: 'itemC',
        isChecked: false,
      },
    },
  },
}

describe('ShopCard', () => {
  it('renders correctly', () => {
    const { container } = render(
      <ShopCard shopId="x" database={testDatabase} />
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  it('renders the correct shop title', () => {
    const { getByText } = render(
      <ShopCard shopId="x" database={testDatabase} />
    )
    const title = getByText('shopX')
    expect(title).toBeInTheDocument()
  })

  it('only renders allocated listItems', () => {
    const { getByText, queryByText } = render(
      <ShopCard shopId="x" database={testDatabase} />
    )
    const itemA = getByText('itemA')
    const itemB = queryByText('itemB')
    const itemC = queryByText('itemC')
    expect(itemA).toBeInTheDocument()
    expect(itemB).not.toBeInTheDocument()
    expect(itemC).not.toBeInTheDocument()
  })

  it('renders the correct summary text', () => {
    const { getByText } = render(
      <ShopCard shopId="x" database={testDatabase} />
    )
    const summaryText = getByText('und 1 abgehakter Eintrag')
    expect(summaryText).toBeInTheDocument()
  })

  fit('links to correct href', () => {
    const historyMock = { push: jest.fn() }
    const { getByTestId } = shallow(
      <ShopCard history={historyMock} shopId="x" database={testDatabase} />
    )

    const card = getByTestId('shop-card')
    console.log({ card })
    user.click(card)
    expect(historyMock.push.mock.calls[0]).toEqual([
      'http://localhost/shop',
      'x',
    ])
    // expect(card).toHaveProperty('href', 'http://localhost/shop')
  })
  /* fit('links to correct href', () => {
    const historyMock = { push: jest.fn() }
    const { getByTestId } = render(
      <Router>
        <ShopCard shopId="x" database={testDatabase} />
      </Router>
    )
    const card = getByTestId('shop-card')
    user.click(card)
    expect(historyMock.push.mock.calls[0]).toEqual([
      'http://localhost/shop',
      'x',
    ])
    // expect(card).toHaveProperty('href', 'http://localhost/shop')
  }) */
})
