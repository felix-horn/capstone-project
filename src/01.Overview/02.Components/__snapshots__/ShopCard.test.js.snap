// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShopCard renders correctly 1`] = `
.c1 {
  font-size: 1.2rem;
  font-weight: 400;
  color: var(--almost-black);
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 5px;
}

.c4 {
  border: 1px solid var(--light-gray);
  border-radius: 2px;
  height: 15px;
  width: 15px;
}

.c5 {
  margin: 0;
  color: var(--dark-gray);
}

.c6 {
  margin: 0;
  font-size: 0.8rem;
  color: var(--light-gray);
}

.c0 {
  border: var(--border);
  border-radius: 8px;
  padding: 15px;
  display: grid;
  gap: 5px;
}

.c2 {
  margin-bottom: 8px;
}

.c7 {
  margin: 4px 2px 0;
}

<div
  class="c0"
  data-testid="shop-card"
>
  <h2
    class="c1 c2"
  >
    shopX
  </h2>
  <label
    class="c3"
  >
    <div
      class="c4"
    />
    <p
      class="c5"
    >
      itemA
    </p>
  </label>
  <p
    class="c6 c7"
  >
    und 1 abgehakter Eintrag
  </p>
</div>
`;
